src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -1618,6 +1620,15 @@ public Solution WithDocumentText(IEnumerable<DocumentId?> documentIds, SourceTex
             : this;
     }
 
+    internal Solution WithSourceGeneratorExecutionVersions(
+        SourceGeneratorExecutionVersionMap sourceGeneratorExecutionVersionMap, CancellationToken cancellationToken)
+    {
+        var newCompilationState = _compilationState.WithSourceGeneratorExecutionVersions(sourceGeneratorExecutionVersionMap, cancellationToken);
+        return newCompilationState != _compilationState
+            ? new Solution(newCompilationState)
+            : this;
```

Comment:
Should there just be a WithCompilationState method?

Reply:
done!